### PR TITLE
Added Linear Regression Model (Scratch)

### PR DESCRIPTION
**Pull request by yashlamba**
Last pull request was automatically closed when I tried to update it with master and piled up all the commits of the repo. (#96)

### CODE REVIEW

File: model/scratch/dffml_model_scratch/model/slr.py [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+    def __init__(self, parent, features):
+        super().__init__(parent, features)
+        self.xData = np.array([])
+        self.yData = np.array([])
+        self.features = self.applicable_features(features)
+
+    def get_regression_line(self):
+        return self.parent.saved.get(hashlib.sha384((''.join(sorted(self.features))).encode()).hexdigest(), None)
+
+    def set_regression_line(self, slope, constant, accuracy):
```
Comment (pdxjohnny) [inline model/scratch/dffml_model_scratch/model/slr.py]: We were planing on doing a `@property` method for `regression_line`

Comment (pdxjohnny) [reply model/scratch/dffml_model_scratch/model/slr.py]:
```python
    @property
    def regression_line(self):
        return self.parent.saved.get(hashlib.sha384((''.join(sorted(self.features))).encode()).hexdigest(), None)

    @regression_line.setter
    def regression_line(self, rline):        
        self.parent.saved[hashlib.sha384((''.join(sorted(self.features))).encode()).hexdigest()] = rline
```

File: model/scratch/dffml_model_scratch/model/slr.py [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+    directory: str  
+
+
+class SLRContext(ModelContext):
+
+    def __init__(self, parent, features):
+        super().__init__(parent, features)
+        self.xData = np.array([])
+        self.yData = np.array([])
+        self.features = self.applicable_features(features)
```
Comment (pdxjohnny) [inline model/scratch/dffml_model_scratch/model/slr.py]:
```python
self._features_hash = hashlib.sha384((''.join(sorted(self.features))).encode()
```

File: model/scratch/dffml_model_scratch/model/slr.py [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+            if feature.dtype() != int and feature.dtype() != float:
+                raise ValueError("Simple Linear Regression only supports int or float feature")
+            if feature.length() != 1:
+                raise ValueError("Simple LR only supports single values (non-matrix / array)")
+            usable.append(feature.NAME)
+        return usable
+
+    async def predict_input(self, x):
+        regression_line = self.get_regression_line()
+        prediction = regression_line[0]*x+regression_line[1]
```
Comment (pdxjohnny) [inline model/scratch/dffml_model_scratch/model/slr.py]:
```python
self.regression_line[0]*x+self.regression_line[1]
```

Now we don't have to call `self.get_regression_line()`

File: model/scratch/dffml_model_scratch/model/slr.py [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+
+    def get_regression_line(self):
+        return self.parent.saved.get(hashlib.sha384((''.join(sorted(self.features))).encode()).hexdigest(), None)
+
+    def set_regression_line(self, slope, constant, accuracy):
+        self.parent.saved[hashlib.sha384((''.join(sorted(self.features))).encode()).hexdigest()] = (slope, constant, accuracy)
+
+    def applicable_features(self, featuress):
+        usable = []
+        if len(featuress) != 1:
```
Comment (pdxjohnny) [inline model/scratch/dffml_model_scratch/model/slr.py]: `featuress` -> `features`

File: model/scratch/dffml_model_scratch/model/slr.py [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+        self.logger.debug("Number of input repos: {}".format(len(self.xData)))
+        x = self.xData
+        y = self.yData
+        mean_x = np.mean(self.xData)
+        mean_y = np.mean(self.yData)
+        m = (mean_x*mean_y - np.mean(x*y))/((mean_x**2) - np.mean(x*x))
+        b = mean_y - (m*mean_x)
+        regression_line = [m*x + b for x in x]
+        accuracy = await self.coeff_of_deter(y, regression_line)
+        self.set_regression_line(m, b, accuracy)
```
Comment (pdxjohnny) [inline model/scratch/dffml_model_scratch/model/slr.py]:
```python
self.regression_line = (m, b, accuracy)
```

File: model/scratch/tests/test_slr.py [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+            for i in range(0, len(Y))
+            ]
+        cls.sources = \
+            Sources(MemorySource(MemorySourceConfig(repos=cls.repos)))
+
+    @classmethod
+    def tearDownClass(cls):
+        cls.model_dir.cleanup()
+
+    async def test_context(self):
```
Comment (pdxjohnny) [inline model/scratch/tests/test_slr.py]: Make sure this is the 3 different, `test_00_train, test_01_accuracy, test_02_predict` methods

File: model/scratch/dffml_model_scratch/model/slr.py [hunk 7]
```diff
@@ -0,0 +1,161 @@
...
+        usable = []
+        if len(featuress) != 1:
+            raise ValueError("Simple Linear Regression doesn't support features other than 1")
+        for feature in featuress:
+            if feature.dtype() != int and feature.dtype() != float:
+                raise ValueError("Simple Linear Regression only supports int or float feature")
+            if feature.length() != 1:
+                raise ValueError("Simple LR only supports single values (non-matrix / array)")
+            usable.append(feature.NAME)
+        return usable
```
Comment (pdxjohnny) [inline model/scratch/dffml_model_scratch/model/slr.py]: don't forget to `sorted(usable)`

File: model/scratch/dffml_model_scratch/model/slr.py [hunk 8]
```diff
@@ -0,0 +1,161 @@
...
+        mean_y = np.mean(self.yData)
+        m = (mean_x*mean_y - np.mean(x*y))/((mean_x**2) - np.mean(x*x))
+        b = mean_y - (m*mean_x)
+        regression_line = [m*x + b for x in x]
+        accuracy = await self.coeff_of_deter(y, regression_line)
+        self.set_regression_line(m, b, accuracy)
+        return (m, b, accuracy)
+
+    async def train(self, sources: Sources):
+        async for repo in sources.with_features(self.features):
```
Comment (pdxjohnny) [inline model/scratch/dffml_model_scratch/model/slr.py]: `async for sources.with_features(self.features + [self.parent.config.predict]):`

**Comment by yashlamba** [comment]
@pdxjohnny I have made all the changes discussed. I'll probably need your help in the README for scratch.

File: model/scratch/dffml_model_scratch/model/slr.py [hunk 9]
```diff
@@ -0,0 +1,160 @@
...
+                                                [self.parent.config.predict]):
+            feature_data = repo.features(self.features +
+                                         [self.parent.config.predict])
+            self.xData = np.append(self.xData, feature_data[self.features[0]])
+            self.yData = np.append(self.yData, feature_data[self.parent.config.predict])
+        self.regression_line = await self.best_fit_line()
+
+    async def accuracy(self, sources: Sources) -> Accuracy:
+        if self.regression_line is None:
+            raise ValueError('Model Not Trained')
```
Comment (pdxjohnny) [inline model/scratch/dffml_model_scratch/model/slr.py]: Will merge this in current state once CI tests pass will change ValueError after #125 is fixed

**Pull request merged**